core/services/pipeline/variables.go

```diff
@@ -0,0 +1,247 @@
+package pipeline
+
+import (
+	"bytes"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+//go:generate mockery --name PipelineParamUnmarshaler --output ./mocks/ --case=underscore
+
+type (
+	Vars map[string]interface{}
+
+	PipelineParamUnmarshaler interface {
+		UnmarshalPipelineParam(val interface{}, vars Vars) error
+	}
+)
+
+var (
+	variableRegexp = regexp.MustCompile(`\$\(([a-zA-Z0-9_\.]+)\)`)
+
+	ErrKeypathNotFound = errors.New("keypath not found")
+)
+
+func NewVars() Vars {
+	return make(Vars)
+}
+
+func (vars Vars) Get(keypath string) (interface{}, error) {
+	parts := keypathParts(keypath)
+	if len(parts) == 0 {
+		return (map[string]interface{})(vars), nil
+	}
+	return vars.traverse(parts, false)
+}
+
+func (vars Vars) Set(keypath string, val interface{}) error {
+	parts := keypathParts(keypath)
+	if len(parts) == 0 {
+		return errors.New("can't set the root of a Vars")
+	}
+
+	last, err := vars.traverse(parts[:len(parts)-1], true)
+	if err != nil {
+		return err
+	}
+	lastKey := parts[len(parts)-1]
+
+	switch typed := last.(type) {
+	case map[string]interface{}:
+		typed[string(lastKey)] = val
+
+	case []interface{}:
+		idx, err := strconv.Atoi(string(lastKey))
+		if err != nil {
+			return err
+		} else if len(typed) <= idx {
+			return errors.New("index out of range")
+		}
+		typed[idx] = val
+
+	default:
+		return errors.New("encountered non-map/non-slice")
+	}
+	return nil
+}
+
+var keypathSeparator = []byte(".")
+
+func keypathParts(keypath string) [][]byte {
+	if len(keypath) == 0 {
+		return nil
+	}
+	// The bytes package uses platform-dependent hardware optimizations and
+	// avoids the extra allocations that are required to work with strings.
+	// Keypaths have to be parsed quite a bit, so let's do it well.
+	kp := []byte(keypath)
+	n := 1 + bytes.Count(kp, keypathSeparator)
+	parts := make([][]byte, n)
+	for i := 0; i < n-1; i++ {
+		nextSep := bytes.IndexByte(kp, keypathSeparator[0])
+		parts[i] = kp[:nextSep]
+		kp = kp[nextSep+1:]
+	}
+	parts[len(parts)-1] = kp
+	return parts
+}
```

Comment:
Is this pulled from some other package?

Reply:
Nope, I wrote it. However, I re-wrote it today, simpler and less flexible now